lib/ide/src/view/text_editor.rs

```diff
@@ -13,25 +16,65 @@ use basegl::display::world::*;
 
 use nalgebra::Vector2;
 use nalgebra::zero;
+use web_sys::KeyboardEvent;
+
 
 
 // ==================
 // === TextEditor ===
 // ==================
 
+shared! { TextEditor
+
 /// TextEditor allows us to edit text files or Enso Modules. Extensible code highlighting is
 /// planned to be implemented for it.
-#[derive(Clone,Debug)]
-pub struct TextEditor {
-    text_field : TextField,
-    padding    : TemporaryPadding,
-    position   : Vector2<f32>,
-    size       : Vector2<f32>
+#[derive(Debug)]
+pub struct TextEditorData {
+    text_field   : TextField,
+    padding      : TemporaryPadding,
+    position     : Vector2<f32>,
+    size         : Vector2<f32>,
+    controller   : Handle,
+    key_listener : Option<KeyboardListener>
 }
 
+impl {
+    /// Updates the underlying display object.
+    pub fn update(&self) {
+        let padding  = self.padding;
+        let position = self.position;
+        let position = Vector3::new(position.x + padding.left, position.y + padding.bottom, 0.0);
+        self.text_field.set_position(position);
+        // TODO: set text field size once the property change will be supported.
+        // let padding  = Vector2::new(padding.left + padding.right, padding.top + padding.bottom);
+        // self.text_field.set_size(self.dimensions - padding);
+        self.text_field.update();
+    }
+
+    /// Saves text editor's content to file.
+    pub fn save(&self) {
+        let controller = self.controller.clone();
+        let mut text = String::new();
+        self.text_field.with_content(|content| {
+            let strings : Vec<String> = content.lines.iter().map(|l| l.chars().iter().collect()).collect();
+            for string in &strings {
+                text += string;
+                text += "\n"
+            }
+            text.pop(); // removes last \n
+        });
+
+        executor::global::spawn(async move {
+            if controller.store_content(text).await.is_err() {
```

Comment:
Don't pass the controller to the async close, we don't need to keep it alive while waiting for the store result.
You can have 
```rust
        let store_fut = controller.store_content(text);
        executor::global::spawn(async move {
            if store_fut.await.is_err() {
                println!("Failed to write file")
            }
        });
```